### ISSUE TITLE
Unbounded growth of disk usage for singleton-blobstore

### ISSUE THREAD

**Issue by ae-govau** [cloudfoundry/cloud_controller_ng#1602]
## Issue

Takes more than a week to delete orphaned blobs

## Context

We deploy (and deploy often) a set of large apps via pipelines. This makes the blobstore grow reasonably quickly.

## Steps to Reproduce

Create lots of big apps.

## Expected result

That a daily (or otherwise) scheduled task would remove orphaned blobs from the blobstore, or that at minimum it be possible for to us to configure (via a BOSH property) a more frequent run of such a task.

## Current result

The cleanup schedule task has non-configurable frequency and runs at 1:30am daily.

And - it has been built so that it cleans up a different section of the blobstore depending on the day of the week.

This make it impossible to run a full cleanup (if needed) without waiting up to a week.

<https://github.com/cloudfoundry/cloud_controller_ng/blob/1ebab1cbb5270a3d51c0a098a37cd9e8ca0f721d/app/jobs/runtime/orphaned_blobs_cleanup.rb#L126>

## Possible Fix

I'll see if I can PR something. This current behaviour is unusual and surprising (most particularly the day of week aspect).

**Comment by ae-govau** [comment]
btw, the reason that's so surprising, is that one generally separates the schedule for a job, from what the job itself does.

The parent class which actually schedules this job (<https://github.com/cloudfoundry/cloud_controller_ng/blob/f330e0aea363d7c1e24f0cd9054966e267b53c31/lib/cloud_controller/clock/scheduler.rb#L16>), which is currently on a daily schedule, ought to quite reasonably make decisions like "change this to a weekly job" without needing to inspect every line of code in a job itself.

By making this depend on the day of week, it's unnecessarily tightly coupling itself to whatever needs to schedule it.

**Comment by selzoc** [comment]
Hello!

The job you've found there is designed to clean up blobs that are orphaned, i.e. that exist in the blobstore and that we don't have a record of in our database.  I'm curious which blobstore bucket you're seeing the growth in?

For example, when you push an app, there is a package of app source, and then when you build the app, you get a droplet, and those are stored in separate buckets.  When you delete the app, the associated blobstore resources are also deleted.

**Comment by ae-govau** [comment]
The day of the week was the first interesting thing I found here... That said, although I really do think that ought be fixed, I don't think it's the root cause of our issues.

What I found was that on singleton-blobstore that the `/var/vcap/store/shared/cc-resources/app_bits_cache` appeared to be growing unbounded and was responsible for 84GB of the 95GB used.

All up that directory was using I think 84GB, 57GB of which hadn't been accessed for at least 7 days (`find /var/vcap/store/shared/cc-resources/app_bits_cache -type f -atime +7 -print0 | du --files0-from=- -sh --total`), 27GB of which had been accessed within 7 days (`find /var/vcap/store/shared/cc-resources/app_bits_cache -type f -atime -7 -print0 | du --files0-from=- -sh --total`)

I noted that this directory is explicitly excluded by the cleanup job.

For this particular environment, we regularly (multiple times a day) push a set of new Java apps to it, then remove the old ones, and have no other applications present.  As such I was quite happy to simply run `find /var/vcap/store/shared/cc-resources/app_bits_cache -type f -atime +7 -delete` to remove the older files, and that freed up some immediate space, however it seems like something isn't cleaning up the way it's meant to.

**Comment by ae-govau** [comment]
I also meant to say, thanks @selzoc for helping us dig into. Also we are running version v12.45.0 of `cf-deployment`.

**Comment by selzoc** [comment]
For historical reasons, the blobs in the `resources` bucket are not automatically cleaned up.  They aren't touched by the `OrphanedBlobsCleanup` job.  That job in particular is designed to be very conservative in its approach to deleting blobs, which is what you have observed.  However, it's not relevant to your particular issue.

I should have linked this document before: https://docs.cloudfoundry.org/concepts/architecture/cloud-controller.html#manual-clean.  It sounds like you've arrived at basically the same solution.

We don't currently have anything planned to clean the `resources` bucket.  One tactical thing you could do that we do support is to disable resource matching via a feature flag: http://v3-apidocs.cloudfoundry.org/version/3.83.0/index.html#list-of-feature-flags.  This would make it so nothing gets placed into the `resources` bucket, but at the cost of having to upload all the files in an app on each `cf push`.

**Comment by ae-govau** [comment]
Thanks @selzoc - so IIUC, the blobstore simply grows unbounded?

Reading the linked doc, is there any good reason to not simply have a daily task that deletes that directory entirely?  If safe, then I'm surprised that someone spent the time to document it, and not the time to write the few lines of code that would do the equivalent as a cron? (perhaps I'm missing something?)

(if not addressable upstream, we'll likely co-locate a daily job that does that on our singleton blobstore, but it sounds like it would be much better handled within `cloud_controller_ng` so as to be applicable across multiple blobstore providers)

**Comment by ae-govau** [comment]
@selzoc - I tried disabling the feature flag as you suggested, however it doesn't seem to stop the files getting stored in the `resources` bucket. All it seems to do is stop the API reporting that the files already exist, so they are uploaded again, and placed in that same directory.

Tested by disabling the flag, deleting the `app_bits_cache`, then re-pushing all apps, and noting the `app_bits_cache` was immediately populated with nearly 2GB of content.

**Comment by sethboyles** [comment]
We certainly shouldn't be uploading resources to the cache if that feature flag is not enabled.  I have added a bug to track that here: https://www.pivotaltracker.com/story/show/172707458.  Thanks for the catch!

**Comment by ae-govau** [comment]
@sethboyles - unclear how one can draw that conclusion from the commit history and public tracker info alone. I can find references to resource matching back to 2015, but I only see this feature flag to disable from late 2019 - and the only [reason given](https://www.pivotaltracker.com/n/projects/966314/stories/152662647):

> As an admin
> I want to be able to disable resource matching
> **So that I can continue to deploy an app when resource matching is causing problems (for instance: a CVE)**

I'll admit I don't understand that reason... but it doesn't say anything about blobstore usage being a goal of that "feature flag".

I mean, we'll take this happily enough as a workaround, but it seems fundamentally wrong that an unbounded cache (which is safe to blow away at any time) - is deliberately being written to a persistent disk. That guarantees long-term failure for every installation that uses it, or, for those using a public cloud provider, a guaranteed recurring cost for storage of bytes no longer needed.

**Comment by ae-govau** [comment]
Since the issue has disappeared in the "Pivotal Tracker" with no updates, here's the workaounnd we've employed on our `cf-deployment`  via an operator file:

```yaml
- type: replace
  path: /instance_groups/name=singleton-blobstore/jobs/-
  value:
    name: pre-start-script
    release: os-conf
    properties:
      script: |
        #!/usr/bin/env bash
        set -euo pipefail

        cat <<EOF > /etc/cron.daily/clear_app_bits_cache
        #!/usr/bin/env bash
        set -euo pipefail
        rm -rf /var/vcap/store/shared/cc-resources/app_bits_cache
        EOF

        chmod a+x /etc/cron.daily/clear_app_bits_cache

        # and run it once now
        /etc/cron.daily/clear_app_bits_cache
```

**Comment by Gerg** [comment]
With d4d891c, the disabling the `resource_matching` feature flag will prevent CC from populating the resource cache.

**Comment by ae-govau** [comment]
Thanks for the update @Gerg

**Comment by Gerg** [comment]
The fix for this issue is included in [capi-release 1.103.0](https://github.com/cloudfoundry/capi-release/releases/tag/1.103.0). Thanks for raising this issue!
